rfcs/0049-flakes.md

```diff
@@ -0,0 +1,708 @@
+
+  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
+  commit hash (`rev`). Note that unlike Git, GitHub allows fetching by
+  commit hash without specifying a branch or tag.
+
+  The only supported parameter is `dir` (see above).
+
+  Some examples:
+
+  * `github:edolstra/dwarffs`
+  * `github:edolstra/dwarffs/unstable`
+  * `github:edolstra/dwarffs/d3f2baba8f425779026c6ec04021b2e927f61e31`
+
+* Indirections through the flake registry. These have the form
+
+      <flake-id>(/<rev-or-ref>(/rev)?)?
+
+  These perform a lookup of `<flake-id>` in the flake registry. The
+  specified `rev` and/or `ref` are then merged with the entry in the
+  registry. (See below.) For example, `nixpkgs` and
+  `nixpkgs/release-19.03` are indirect flake references.
+
+In the future, we should also add tarball flake references
+(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs#hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs#hello
+
+There are multiple registries:
+
+* The global registry
+  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
+  location can be overriden via the `flake-registry` option.) Nix
+  automatically fetches this registry periodically. The check interval
+  is determined by the `tarball-ttl` option.
+
+* The local registry `$XDG_CONFIG_HOME/nix/registry.json`.
+
+* Registry entries specified on the command line.
+
+A registry is a JSON file that looks like this:
+
+```
+{
+    "version": 1,
+    "flakes": {
+        "nixpkgs": {
+            "uri": "github:NixOS/nixpkgs"
+        },
+        ...
+    }
+}
+```
+
+With this registry, flake references resolve as follows:
+
+* `nixpkgs` -> `github:NixOS/nixpkgs`
+* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
+* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`
+
+The registries are searched in reverse order. Thus the local registry
+overrides the global registry, and the command line takes precedence
+over the local and global registries.
+
+## `nix` command line interface
+
+I propose to make flakes the primary way to specify packages in the
+`nix` command line interface. Note that the `nix` UI is still marked
+as experimental, so we have some freedom to make incompatible
+changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
+`nix-instantiate`) should not be changed to avoid breakage.
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs#hello dwarffs#dwarffs
```

Comment:
The hash mark here seems bad! Can't we use something like ```.``` that is already in Nix syntax? Otherwise we end up making everyone quote all their command line args, due to shell using ```#``` as a comment syntax.

Reply:
No need to quote. It needs to be separated by a space first

```
~ $ echo hi#there
hi#there
```

Spaces in URLs will be encoded ( `%20`)

Reply:
Well, in ZSH quoting is still needed.

Reply:
```
clancywiggum% echo $ZSH_VERSION
5.7.1
clancywiggum% echo hi#there
hi#there
clancywiggum% echo it#does#not
it#does#not

```

Reply:
The separator used to be a colon (`:`) until a week ago or so, when I changed it a hash because it's less ambiguous and fits in a bit nicer with the URL-like syntax. The reason why I didn't use the hash from the start was that like Matthew I assumed incorrectly that it would be interpreted as a comment.